# Libraries

In [1]:
!pip install hazm
!pip install tensorflow==2.13.0
!pip install numpy==1.25.0

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 978.8 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 892.6/892.6 kB 21.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.3/17.3 MB 29.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.4/4.4 MB 34.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.7/26.7 MB 33.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 243.3/243.3 kB 19.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 15.1 MB/s eta 0:00:00
  Created wheel for flashtext: filename=flashtext-2.7-py2.py3-none-any.whl size=9300 sha256=8ac9c7950e0bbed7274bd7ff3a271e9e1fecde9e2308d4be4dd1795d4a48f1fd
  Stored in directory: /root/.cache/pip/wheels/49/20/47/f03dfa8a7239c54cbc44ff7389eefbf888d2c1873edaaec888
Successfully built flashtext
  Attempting uninstall: numpy
   

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 524.2/524.2 MB 2.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 51.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.9/294.9 kB 26.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 85.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 440.8/440.8 kB 27.3 MB/s eta 0:00:00
  Attempting uninstall: typing-extensions
    Found existing installation: typing_extensions 4.13.2
    Uninstalling typing_extensions-4.13.2:
      Successfully uninstalled typing_extensions-4.13.2
  Attempting uninstall: protobuf
    Found existing installation: protobuf 5.29.4
    Uninstalling protobuf-5.29.4:
      Successfully uninstalled protobuf-5.29.4
  Attempting uninstall: keras
    Found existing installation: keras 3.8.0
    Uninstalling keras-3.8.0:
      Successfully uninstalled keras-3.8.0
  Attempting uninstall: gast
    Found existing installation: gast 0.6.0
    Uninstalling gast

In [1]:
!pip install datasets transformers[torch] evaluate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 55.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 37.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 52.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 2.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 69.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.8/45.8 kB 3.8 MB/s eta 0:00:00
  Attempting uninstall: typing-extensions
    Fo

In [2]:
import pandas as pd
import numpy as np
import tensorflow as tf
from hazm import Normalizer, WordTokenizer, stopwords_list
import string
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score, confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import MultinomialNB
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.metrics import Precision, Recall, F1Score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout
from datasets import Dataset, DatasetDict, load_dataset
from transformers import AutoTokenizer, DataCollatorWithPadding
from transformers import AutoModelForSequenceClassification
from transformers import Trainer, TrainingArguments
import evaluate

# Data Preprocess

## Downloading Dataset

In [3]:
splits = {'train': 'train.csv', 'validation': 'validation.csv', 'test': 'test.csv'}
train = pd.read_csv("hf://datasets/PNLPhub/snappfood-sentiment-analysis/" + splits["train"])
val = pd.read_csv("hf://datasets/PNLPhub/snappfood-sentiment-analysis/" + splits["validation"])
test = pd.read_csv("hf://datasets/PNLPhub/snappfood-sentiment-analysis/" + splits["test"])

print("training set:", train.shape)
print("validation set:", val.shape)
print("test set:", test.shape)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


training set: (52110, 3)
validation set: (8337, 3)
test set: (9033, 3)


## Quality Evaluation

In [4]:
def clean_dataset(data):
  data = data.drop(labels=['label'], axis=1)
  data = data.dropna()
  return data

train = clean_dataset(train)
val = clean_dataset(val)
test = clean_dataset(test)

print("training set:", train.shape)
print("validation set:", val.shape)
print("test set:", test.shape)

training set: (52110, 2)
validation set: (8337, 2)
test set: (9033, 2)


In [5]:
positive, negative = train['label_id'].value_counts()
print("Number of positive reviews:", positive)
print("Number of negative reviews:", negative)

Number of positive reviews: 26236
Number of negative reviews: 25874


## Normalization

In [6]:
normalizer = Normalizer()

def text_normalization(text):
  return normalizer.normalize(text)

train['normalized'] = train['comment'].apply(lambda x: text_normalization(x))
val['normalized'] = val['comment'].apply(lambda x: text_normalization(x))
test['normalized'] = test['comment'].apply(lambda x: text_normalization(x))

train.to_csv('train.csv', index=False)
val.to_csv('val.csv', index=False)
test.to_csv('test.csv', index=False)

train.head()

comment  label_id  \
0   غذا خیلی سرد بود در صورتیکه فاصله ما خیلی کم است       1.0   
1     بهتره بتونیم ران یا سینه رو خودمون انتخاب کنیم       0.0   
2  غذا بد بود حالم خیییییلی بده. دل دردو دل پیچه....       1.0   
3  با سلام سابق بر این بسته بندی از کیفیت بهتری ب...       1.0   
4                          سلام، خیلی ممنون و متشکرم       0.0   

                                          normalized  
0   غذا خیلی سرد بود در صورتیکه فاصله ما خیلی کم است  
1     بهتره بتونیم ران یا سینه رو خودمون انتخاب کنیم  
2  غذا بد بود حالم خیلی بده. دل دردو دل‌پیچه. معل...  
3  با سلام سابق بر این بسته بندی از کیفیت بهتری ب...  
4                          سلام، خیلی ممنون و متشکرم

# TF-IDF

## Creating TF-IDF Table

In [ ]:
tokenizer = WordTokenizer()
sw = stopwords_list()

def clean_data(text):
  text_no_punc = re.sub(f'[{string.punctuation}؟،٪×÷»«]+', '', text)
  tokens = tokenizer.tokenize(text_no_punc)
  return [token for token in tokens if token not in sw]

tfidf_v = TfidfVectorizer(analyzer=clean_data)
X = tfidf_v.fit_transform(pd.concat([train['normalized'], test['normalized']]))
X_features = pd.DataFrame(X.toarray())
X_train = X_features.head(train.shape[0])
X_test = X_features.tail(test.shape[0])

print(X_train.shape)
print(X_test.shape)

(52110, 31141)
(9033, 31141)


## Logistic Regression

In [ ]:
lr = LogisticRegression()
lr_model = lr.fit(X_train, train['label_id'])

In [ ]:
y_pred = lr_model.predict(X_test)
accuracy = accuracy_score(test['label_id'], y_pred)
precision = precision_score(test['label_id'], y_pred, pos_label=0)
recall = recall_score(test['label_id'], y_pred, pos_label=0)
f1 = f1_score(test['label_id'], y_pred, pos_label=0)
print("Accuray:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)
print("Confusion Matrix:\n", confusion_matrix(test['label_id'], y_pred))

Accuray: 0.8207682940329901
Precision: 0.8357929257437339
Recall: 0.7953633526526973
F1 Score: 0.8150770988006854
Confusion Matrix:
 [[3568  918]
 [ 701 3846]]


## Random Forrest

In [ ]:
rf = RandomForestClassifier(max_depth=2, random_state=0)
rf_model = rf.fit(X_train, train['label_id'])

In [ ]:
y_pred = rf_model.predict(X_test)
accuracy = accuracy_score(test['label_id'], y_pred)
precision = precision_score(test['label_id'], y_pred, pos_label=0)
recall = recall_score(test['label_id'], y_pred, pos_label=0)
f1 = f1_score(test['label_id'], y_pred, pos_label=0)
print("Accuray:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)
print("Confusion Matrix:\n", confusion_matrix(test['label_id'], y_pred))

Accuray: 0.6901361673862504
Precision: 0.6365549619556419
Recall: 0.8765046812304949
F1 Score: 0.7375035168339117
Confusion Matrix:
 [[3932  554]
 [2245 2302]]


## MultinomialNB

In [ ]:
mnb = MultinomialNB()
mnb_model = mnb.fit(X_train, train['label_id'])

In [ ]:
y_pred = mnb_model.predict(X_test)
accuracy = accuracy_score(test['label_id'], y_pred)
precision = precision_score(test['label_id'], y_pred, pos_label=0)
recall = recall_score(test['label_id'], y_pred, pos_label=0)
f1 = f1_score(test['label_id'], y_pred, pos_label=0)
print("Accuray:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)
print("Confusion Matrix:\n", confusion_matrix(test['label_id'], y_pred))

Accuray: 0.806376619063434
Precision: 0.8646416200373035
Recall: 0.7233615693267945
F1 Score: 0.787716955941255
Confusion Matrix:
 [[3245 1241]
 [ 508 4039]]


# LSTM

## Tokenization

In [ ]:
max_words = 32000
# defining a tokenizer
tokenizer = Tokenizer(num_words=max_words)
tokenizer.fit_on_texts(train['normalized'])

In [ ]:
# convert text to sequence of numbers
train_seq = tokenizer.texts_to_sequences(train['normalized'])
val_seq = tokenizer.texts_to_sequences(val['normalized'])
test_seq = tokenizer.texts_to_sequences(test['normalized'])

In [ ]:
# padding sequences
train_seq_padded = pad_sequences(train_seq)
val_seq_padded = pad_sequences(val_seq)
test_seq_padded = pad_sequences(test_seq)

## Training

In [ ]:
# defining model
model = Sequential()
model.add(Embedding(input_dim=max_words, output_dim=50))
model.add(LSTM(32, dropout=0, recurrent_dropout=0))
model.add(Dense(32, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [ ]:
# training model
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy', Precision(), Recall(), F1Score()])
history = model.fit(train_seq_padded, train['label_id'],
                    batch_size=256, epochs=10,
                    validation_data=(val_seq_padded, val['label_id']))

Epoch 1/10
204/204 [==============================] - 122s 582ms/step - loss: 0.4363 - accuracy: 0.7999 - precision: 0.7872 - recall: 0.8182 - f1_score: 0.6636 - val_loss: 0.3480 - val_accuracy: 0.8557 - val_precision: 0.8255 - val_recall: 0.8998 - val_f1_score: 0.6639
Epoch 2/10
204/204 [==============================] - 109s 534ms/step - loss: 0.3073 - accuracy: 0.8757 - precision: 0.8465 - recall: 0.9157 - f1_score: 0.6636 - val_loss: 0.3618 - val_accuracy: 0.8407 - val_precision: 0.8537 - val_recall: 0.8199 - val_f1_score: 0.6639
Epoch 3/10
204/204 [==============================] - 113s 554ms/step - loss: 0.2662 - accuracy: 0.8945 - precision: 0.8703 - recall: 0.9255 - f1_score: 0.6636 - val_loss: 0.3553 - val_accuracy: 0.8474 - val_precision: 0.8149 - val_recall: 0.8967 - val_f1_score: 0.6639
Epoch 4/10
204/204 [==============================] - 110s 541ms/step - loss: 0.2355 - accuracy: 0.9085 - precision: 0.8874 - recall: 0.9342 - f1_score: 0.6636 - val_loss: 0.3905 - val_accur

In [ ]:
# saving model
model.save('lstm.keras')

## Model Validation

In [ ]:
# calculating metrics on test set
y_pred = np.round(model.predict(test_seq_padded))
accuracy = accuracy_score(test['label_id'], y_pred)
precision = precision_score(test['label_id'], y_pred, pos_label=0)
recall = recall_score(test['label_id'], y_pred, pos_label=0)
f1 = f1_score(test['label_id'], y_pred, pos_label=0)
print("Accuray:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)
print("Confusion Matrix:\n", confusion_matrix(test['label_id'], y_pred))

283/283 [==============================] - 7s 22ms/step
Accuray: 0.8209897044171371
Precision: 0.8347724620770128
Recall: 0.7973695942933571
F1 Score: 0.8156424581005587
Confusion Matrix:
 [[3577  909]
 [ 708 3839]]


# ParsBERT

In [7]:
# creating datasets
train['label_id'] = train['label_id'].astype(int)
val['label_id'] = val['label_id'].astype(int)
test['label_id'] = test['label_id'].astype(int)
train_set = Dataset.from_pandas(train.drop(labels=['comment'], axis=1))
val_set = Dataset.from_pandas(val.drop(labels=['comment'], axis=1))
test_set = Dataset.from_pandas(test.drop(labels=['comment'], axis=1))
train_set = train_set.rename_column("label_id", "labels")
train_set = train_set.rename_column("normalized", "text")
val_set = val_set.rename_column("label_id", "labels")
val_set = val_set.rename_column("normalized", "text")
test_set = test_set.rename_column("label_id", "labels")
test_set = test_set.rename_column("normalized", "text")
print(train_set)
print(val_set)
print(test_set)

Dataset({
    features: ['labels', 'text'],
    num_rows: 52110
})
Dataset({
    features: ['labels', 'text'],
    num_rows: 8337
})
Dataset({
    features: ['labels', 'text'],
    num_rows: 9033
})


In [13]:
# loading model
checkpoint = "HooshvareLab/bert-base-parsbert-uncased"

# using parsbert tokenizer
tokenizer = AutoTokenizer.from_pretrained(checkpoint, model_max_length=256)
def tokenize_function(example):
    return tokenizer(example["text"], truncation=True)

tokenized_train_set = train_set.map(tokenize_function, batched=True)
tokenized_val_set = val_set.map(tokenize_function, batched=True)
tokenized_test_set = test_set.map(tokenize_function, batched=True)
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

Map:   0%|          | 0/52110 [00:00<?, ? examples/s]

Map:   0%|          | 0/8337 [00:00<?, ? examples/s]

Map:   0%|          | 0/9033 [00:00<?, ? examples/s]

In [14]:
# training
training_args = TrainingArguments(
    "snappfood-trainer",
    eval_strategy="epoch",
    report_to="none",
    save_strategy="epoch",
    overwrite_output_dir=True,
    num_train_epochs=2.0
)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)

def compute_metrics(eval_preds):
    precision = evaluate.load("precision")
    recall = evaluate.load("recall")
    accuracy = evaluate.load("accuracy")
    f1 = evaluate.load("f1")

    logits, labels = eval_preds
    predictions = np.argmax(logits, axis=-1)
    return {'precision': precision.compute(predictions=predictions, references=labels)['precision'],
            'recall': recall.compute(predictions=predictions, references=labels)['recall'],
            'accuracy': accuracy.compute(predictions=predictions, references=labels)['accuracy'],
            'f1': f1.compute(predictions=predictions, references=labels)['f1']}

trainer = Trainer(
    model,
    training_args,
    train_dataset=tokenized_train_set,
    eval_dataset=tokenized_val_set,
    data_collator=data_collator,
    processing_class=tokenizer,
    compute_metrics=compute_metrics
)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at HooshvareLab/bert-base-parsbert-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [15]:
trainer.train()

TrainOutput(global_step=13028, training_loss=0.40258062804517547, metrics={'train_runtime': 1799.0582, 'train_samples_per_second': 57.93, 'train_steps_per_second': 7.242, 'total_flos': 2824284429334680.0, 'train_loss': 0.40258062804517547, 'epoch': 2.0})

## Model Validation

In [16]:
raw_pred, label_ids, metrics = trainer.predict(tokenized_test_set)
y_pred = np.argmax(raw_pred, axis=1)

In [17]:
# calculating metrics on test set
accuracy = accuracy_score(test['label_id'], y_pred)
precision = precision_score(test['label_id'], y_pred, pos_label=0)
recall = recall_score(test['label_id'], y_pred, pos_label=0)
f1 = f1_score(test['label_id'], y_pred, pos_label=0)
print("Accuray:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)
print("Confusion Matrix:\n", confusion_matrix(test['label_id'], y_pred))

Accuray: 0.8633897929812908
Precision: 0.9069069069069069
Recall: 0.8078466339723585
F1 Score: 0.8545154444706438
Confusion Matrix:
 [[3624  862]
 [ 372 4175]]


# Comparison

As you can see in the table  below, **ParsBert** outperforms all the other models in accuracy, precision and F1 score.

| Metric      | TF-IDF (Logistic Regression) | TF-IDF (Random Forrest) | TF-IDF (MultinomialNB)| LSTM| ParsBert |
| ----------- | -----------                  |-----------                  |-----------   | ------- | ----- |
| Accuracy      |82.08%                    |69.01%                        |80.64% |82.10% |**86.34%** |
| Precision   |83.58%                          |63.66%                      |86.46% |83.48% |**90.69%** |
| Recall   |79.54%                          |**87.65%**                         |72.34% |79.74% |80.78% |
| F1 Score   |81.51%                      |73.75%                         |78.77% |81.56% |**85.45%** |